# DataFrames Exercises

Do your work for this exercise in a python script or a jupyter notebook with the name dataframes.py or dataframes.ipynb.

For several of the following exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.)


            from pydataset import data


When the instructions say to load a dataset, you can pass the name of the dataset as a string to the data function to load the dataset. You can also view the documentation for the data set by passing the show_doc keyword argument.


       # data('mpg', show_doc=True) # view the documentation for the dataset
    mpg = data('mpg') # load the dataset and store it in a variable
    
    
All the datasets loaded from the pydataset library will be pandas dataframe

In [12]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

In [15]:
print(df)

       name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72


In [28]:
df.sample(3)

name  math  english  reading
4   Ada    77       92       98
1  Jane    88       79       67
5  John    79       76       93

In [59]:
english_passing = df[df.english > 70]

In [60]:
english_passing

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

1). Copy the code from the lesson to create a dataframe full of student grades.

a). Create a column named passing_english that indicates whether each student has a passing grade in english.

In [36]:
df['passing_english'] = df.english > 70


In [42]:
df = df.drop(columns=['passing_grade'])

b). Sort the english grades by the passing_english column. How are duplicates handled?



In [45]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

c). Sort the english grades first by passing_english and then by student name. All the students 
that are failing english should be first, and within the students that are failing english 
they should be ordered alphabetically. The same should be true for the students passing 
english. (Hint: you can pass a list to the .sort_values method)


In [49]:
df.sort_values(by='passing_english').sort_values(by='name')

name  math  english  reading  passing_english
4       Ada    77       92       98             True
11     Alan    92       62       72            False
8    Albert    92       62       87            False
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
7     Marie    93       63       90            False
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81            False

d). Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.



In [55]:
df.sort_values(by=['passing_english']).sort_values(by=['english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

e). Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.



In [140]:
df['overall_grade'] = df.mean(axis = 1)

/var/folders/b2/3fdswr7x4v5fhdt54wpcfmdw0000gn/T/ipykernel_935/628609030.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df['overall_grade'] = df.mean(axis = 1)


In [141]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True          57.00
1      Jane    88       79       67             True          58.75
2     Suzie    94       74       95             True          66.00
3     Billy    98       96       88             True          70.75
4       Ada    77       92       98             True          67.00
5      John    79       76       93             True          62.25
6    Thomas    82       64       81            False          56.75
7     Marie    93       63       90            False          61.50
8    Albert    92       62       87            False          60.25
9   Richard    69       80       94             True          61.00
10    Isaac    92       99       93             True          71.25
11     Alan    92       62       72            False          56.50

# 2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:



In [250]:
from pydataset import data
mpg = data('mpg')

In [251]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> How many rows and columns are there?
    
    Answer: 234 rows × 11 columns

In [227]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

    --> What are the data types of each column?

In [252]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

    --> Summarize the dataframe with .info and .describe

In [253]:
mpg.info


<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

In [254]:
mpg.describe

<bound method NDFrame.describe of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    

    --> Rename the cty column to city.

In [255]:
mpg = mpg.rename(columns={'cty':'city'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> Rename the hwy column to highway.

In [256]:
mpg = mpg.rename(columns={'hwy':'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

    --> Do any cars have better city mileage than highway mileage?
    
    Answer: 0

In [257]:
mpg[mpg.city > mpg.highway]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

    --> Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [258]:
mpg['mileage_difference'] = np.where((mpg['city'] >= mpg['highway']) & (mpg['city'] <= mpg['highway']), True, False)

mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact               False  
2    compact               False  
3    compact               False  
4    compact               False  
5    compact               False  
..       ...                 ...  
230  midsize               False  
231  midsize               False  
232  midsize               False  
233  midsize               False  
234  midsize               False  

[234 rows x 12 columns]

In [268]:
conditions = [mpg['city'] > mpg['highway'], mpg['city'] < mpg['highway'], mpg['city'] == mpg['highway']]

choices = ['city > highway', 'highway > city', 'city == highway']

mpg['mileage_difference'] = np.select(conditions, choices)

mpg


manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class mileage_difference  
1    compact     highway > city  
2    compact     highway > city  
3    compact     highway > city  
4    compact     highway > city  
5    compact     highway > city  
..       ...                ...  
230  midsize     highway > city  
231  midsize     highway > city  
232  midsize     highway > city  
233  midsize     highway > city  
234  midsize     highway > city  

[234 rows x 12 columns]

    --> Which car (or cars) has the highest mileage difference?

TypeError: min expected at least 1 argument, got 0

    --> Which compact class car has the lowest highway mileage? The best?

In [346]:

# mpg[mpg.city if [mpg.class] == 'compact']

# mpg[mpg.city < 10 ].sort_values(by='class')

# mpg[mpg.city].min().sort_values(by='compact')

1      18
2      21
3      20
4      21
5      16
       ..
230    19
231    21
232    16
233    18
234    17
Name: city, Length: 234, dtype: int64

    --> Create a column named average_mileage that is the mean of the city and highway mileage.

    --> Which dodge car has the best average mileage? The worst?

# 3. Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

    --> How many rows and columns are there?

    --> What are the data types?

    --> Summarize the dataframe with .info and .describe

    --> What is the the weight of the fastest animal?

    --> What is the overal percentage of specials?

    --> How many animals are hoppers that are above the median speed? What percentage is this?